### ISSUE TITLE
Thoughts

### ISSUE THREAD

**Issue by fasterthanlime** [fasterthanlime/cork#1]
Hey @shamanas, @zhaihj - what do you think of the various files hanging in `docs/` ? Are those good ideas or am I just delirious from too much rock bugfixing ?

**Comment by alexnask** [comment]
I read through them earlier, I think everything mentioned up to now is well thought out and reasonable, although I will say I am a bit sad about ditching the poor pluggable frontend and backend idea :P  

I will be reading through them again to see if I can come up with some insightful feedback on details

**Comment by fasterthanlime** [comment]
> ditching the poor pluggable frontend and backend idea

Wellll if only it wasn't such a pain to build cross-platform dynamic libraries :( If coded carefully, I'd say the frontends & backends could be made pluggable later on.

The thing currently on my mind is passes, what exactly should be done in various passes - for example, everything could be resolved with tuple-decls and tuple-assigns, but then they could be unwrapped in a later phase (when translated to another AST?) so the backend's job is easier..

Ideally the middle-end would be written as if the backend could be anything (LLVM, JavaScript, the JVM, who knows) - but it's going to be hard as we've started to integrate some pretty C-specific things in rock (like type promotion rules).

**Comment by alexnask** [comment]
Actually, lightweight resumable tasks sound like a brilliant idea and I think they fit the way I think of the resolving phase perfectly (jumping into a node, getting the info you need, resuming in the parent node, etc..)  

I think decoupling the ooc and C AST is also one of the things we need the most, it really is confusing at the moment and, as you just mentioned, the middle-end should be "back-end agnostic" (for lack of a better term at the moment)  

Figuring out what passes are required and their order in order to eliminate looping will be really tricky but I'm sure we can figure it out.

**Comment by alexnask** [comment]
The way FFI is done is the biggest obstacle for drastically different backends IMHO (although LLVM should be fairly easy), but I don't exactly have a better idea in mind

**Comment by vendethiel** [comment]
> Ideally the middle-end would be written as if the backend could be anything (LLVM, JavaScript, the JVM, who knows) - but it's going to be hard as we've started to integrate some pretty C-specific things in rock (like type promotion rules).

Isn't that the backend?

**Comment by alexnask** [comment]
More specialized AST nodes will also be really nice, the most confusing example I can think of is the way ACS closures are handled in FunctionDecl, I always had a hard time finding what code is ACS-specific and keeping the possibility of the decl being a closure in mind while fixing or implementing things.

**Comment by fasterthanlime** [comment]
> I think decoupling the ooc and C AST is also one of the things we need the most, it really is confusing at the moment and, as you just mentioned, the middle-end should be "back-end agnostic" (for lack of a better term at the moment)

Yep, that's why `libovum` is in there (copied from oc). I don't remember exactly how functional it is, but I think it was a fork of libcee.

> Actually, lightweight resumable tasks sound like a brilliant idea and I think they fit the way I think of the resolving phase perfectly (jumping into a node, getting the info you need, resuming in the parent node, etc..)

For the `resolve` pass (for lack of a better name.. finding the refs of everything (incl. operator overloads), inferring types, etc. - but not closure unwrapping for example), imagine this: the resolver has a list of Trails, but instead of being just Stacks, Trails have a node, and a parent Trail (potentially null).

If we respect the constraint that the resolve pass should only be straightforward "replace ourselves" and nothing more (no inserts), then the interesting part is that whenever a node is replaced by something else, its trail is changed accordingly, and then _all the children trails_ are adjusted automatically (since they all refer to that trail).

Take this code:

``` ooc
msg := "hello" + " world"
```

For the `"hello"` node, we'd have something like:

```
Trail(stringliteral, Trail(binaryop, Trail(vdfe, Trail(module, null))))
```

In this case, the binaryop is overloaded, so it'll be replaced by a functioncall, so:

```
Trail(binaryop, ...)
```

Will be modified to:

```
Trail(functioncall, ...)
```

And the stringliteral trail remains valid!

```
Trail(stringliteral, Trail(functioncall, Trail(vdfe, Trail(module, null))))
```

(In this case it's kinda moot because stringliteral wasn't waiting on anything in the first place.. but you get the idea).

So, the resolver would have a bunch of Trail, some deeper than others, and it'll know what trails are waiting on what other trails, so it can resume their resolving whenever it's ready. Or, even if it doesn't know what it's waiting for, it can loop through unresolved trails, but that's much better than traversing the whole AST (and the common parts of trails is shared).

Not sure if my explanation is good enough, I'm quite tired.

**Comment by fasterthanlime** [comment]
> Isn't that the backend?

Unfortunately, no. When you do:

```
a := (42 as Int) + (13 as LLong)
```

rock has to infer the type of `a` somehow, and it's now using C rules (as of 0.9.10). So it's still done with C in mind, even in the middle end.

**Comment by fasterthanlime** [comment]
> More specialized AST nodes will also be really nice, the most confusing example I can think of is the way ACS closures are handled in FunctionDecl, I always had a hard time finding what code is ACS-specific and keeping the possibility of the decl being a closure in mind while fixing or implementing things.

Definitely. I think ACS unwrapping should be a separate pass (probably specific to the C backend).

To clarify: ACS argument types and return type inference should happen during `resolve`, so that it can behave like a regular functiondecl - but then, splitting it into a thunk & context, that's the backend's job.

**Comment by vendethiel** [comment]
Oh, okay. I guess itd be better in a "perfect world" to carry out the same semantics everywhere, but probably not easy to deal with. It'd certainly help with interop, though ("write once, change integer semantics everywhere kind of style")

**Comment by fasterthanlime** [comment]
> Oh, okay. I guess itd be better in a "perfect world" to carry out the same semantics everywhere, but probably not easy to deal with. It'd certainly help with interop, though ("write once, change integer semantics everywhere kind of style")

Well, good luck emulating C integer semantics in something like JavaScript.. it would be doable in most VMs. (I don't know why I have JavaScript in mind, I don't even really want to compile ooc to JS... it'd just be fun I guess).

**Comment by vendethiel** [comment]
targetting LLVM gives you JS compilation "almost for free", though, with emscripten?

**Comment by fasterthanlime** [comment]
@vendethiel Well... I would call what emscripten does 'compiling to a VM written in JavaScript' rather than 'compiling to JavaScript'.

Besides, you can already compile the C code generated by rock through emscripten, if that's what your heart desires :)

**Comment by vendethiel** [comment]
Right. I guess you're sacrificing a few optimizations, though :).

**Comment by alexnask** [comment]
@fasterthanlime  
Don't worry, you're making perfect sense and I think this is a great way of doing things.
It is also a way that makes parallelism way easier to implement, if it is ever desired.  

@vendethiel You can also run emscripten on the C sources, although IIRC there needed to be a couple of hacks, [here](https://github.com/shamanas/ooc-js) is a (really old) modified version of rock that does it :p

**Comment by alexnask** [comment]
Oops, it seems the emscripten branch of my rock fork is gone forever (I think I deleted my fork and re-forked at some point for whatever reason), too bad :P

**Comment by horasal** [comment]
I'm really happy to see a new start of ooc-compiler.
I read `doc/*` this morning. Everything seems good and reasonable. More passes and types of ASTs are definitely necessary. I'm also expecting a `parent` variable for each node so resolver can know where it is without searching the trail (Trail in Trail also sounds good). 

Immutable AST will make compiler much faster. In fact, I'm wondering if the replace-process should be treated as backend's work. Taking `tuple` for example, unwrapping it to `VariableDecl..., BinaryOp..., VariableAccess` is clearly a c-based action. If someone is making `go` or `python` backend, there's even no need to touch the tuple itself.

As is mentioned previously, separating backend and middle-end may be a challenge. Compiler have to infer types in middle-end, however, priorities of types are backend-dependent. I don't know how to do it neatly. 

Additionally, here're some other points I've been considering. (maybe too early to discuss them, though)
- Keep gc-free in mind

Closure, oocarray, generic... Too many things turn into `gc_malloc` in rock. The new compiler should keep gc-free in mind. Eventhough it may be impossible to free them automatically, at least there should be a way for user to free them manually.
- `Template`

I won't say rock's generic is bad, but it is a big source of bugs. `memcpy` is inefficient (most of time `T` is just something like `Int` which can be stored in regs..), `gc_malloc` leaks under `gc=off`, many modules should judge if a node contains generics or not. I would like to see a c++-like template happens in the new compiler.
- oocarray

Array is still a pain in rock. `Int[]` becomes `ooc_array` while `Int[4]` is `int[4]`. As OOC has generic class, treating `T[]` as `Array<T>` from the very beginning is not a bad idea for me.

**Comment by davidhesselbom** [comment]
I know I wasn't asked, but you seem to be having so much fun, so:

> Keep gc-free in mind

I've been told several times that rock wasn't designed for speed (or something to that effect), but if cork has any such ambitions, perhaps it could allow for stack allocation of objects as in C++, where, as I understand it, stack allocated objects still call their destructor when they go out of scope. I don't know what kind of performance penalties checking the stack for objects to destroy when leaving a scope would incur, but it should be faster than heap allocation and deallocation for every object, at least (or else I don't see why C++ would do it).

The codebase I'm working on is littered with `free()` calls because we turned off the GC (which, again, we've been told we shouldn't), and strings are a nightmare to keep from leaking memory. If cork can be less liberal with calling gc_malloc (when dealing with e.g. generics, as mentioned above and before) and if stack allocation were possible, turning off the GC might not be such a problem in the first place.

I should add I expect nothing, so don't get angry because I suggested something that's really hard to do.

**Comment by alexnask** [comment]
@zhaihj  

`T[] -> Array<T>` makes sense and I think it was actually planned when template covers where implemented (defining Array 100% in ooc with template covers).  

IMHO template covers are sufficient for template programming and reified generics offer lots of advantages.  

I agree 100% with more gc-free support, although it will be a difficult task and if there is no additional feature planned for memory management (RAII, a Rust-like lifetime system or "just" smart+shared pointers), I don't think it is worth it.  

As @davidhesselbom said, code just devolves into a new-free mess, which is amplified due to the inability to allocate objects on the stack (which, tbh, would be a pretty easy thing to implement if we can settle on some syntax and we all agree they are useful, although it is an additional tie to the C-only backend).  

I'm not too sure how RAII would be implemented in the C backend (the only way I can think of is manually adding appropriate destructor calls before every return statement but this still leaves issues with Exceptions, especially since they are conventional function calls and I doubt C++ approaches the problem this way)  

I think it's a good time to agree on any features we want to see supported and removed and make breaking changes.  

(My personal favorite would be adding some kind of unique and shared pointer type and slowly moving away from GC-only)

**Comment by horasal** [comment]
I will be satisfied enough if I can manage memory manually, just like in c. In rock, the real problem is that there's even no way to free generics or closures manually. 

For generics, I'm sorry that I still can't understand the advantages of current generic system. 
Evenything I can imagine works with template. Some viewpoints that why I support template system:

With template:
- No `gc_malloc` and `memcpy` anymore.
  generics will no longer be pointers but real types. There's no need to alloc memory and move the data around. It can be also optimized by c-compiler more efficiently (pass through registers, caching...).
- Match type statically
  `match(T)` or similar codes can be evaluated during compilation time, just like `static if` in D.
- decoupling between resolvers
  In rock, almost every module (including backend) should check if a variable is generic or not. With template, generics will just act as normal types so there's no need to do that type of check anymore.
- Easier to handle typeargs
  issue 804, 806, 830, 822, 415 are all related to generic. Some of them are cause by that rock is not doing enough check (even though there are already many codes related to generic). Some others are not easy to be fixed just because typeargs do not have enough information during compilation time.
  Such as `extend ArrayList<T>` and `extend ArrayList<Int>`, it is expected that `extend ArrayList<T>` adds function for all types and `extend ArrayList<Int>` only works for `Int`. I don't know how to achieve it with current generic system. But with template, what we need to do is simply search for the realized `ArrayList` and `functions addAll`. (For `extend ArrayList<T>`, it is a little difficult because we don't know if T is a user-defined type or generics, so something like `extend<T> ArrayList<T>` can be introduced)

I've been told several times that rock won't change its design of generic. I guessed changing may need too much refactoring. If cork also implement the same style of generic with rock, I would like to know the advantages of it so at least I can use it efficiently. 

Just a one-line (actually four-lines) story: 
I wrote a graph mining program several months ago, mainly with `ArrayList`, but it is 10x slower than c-version when processing graph contains more than 10 thousands vertices. The reason I finally figured out is that `gc_malloc` and `memcpy` takes 70% of total runtimes - almost of them are happened in `ArrayList get`.

**Comment by fasterthanlime** [comment]
High-level notes on the above:
- I don't want cork to turn into 'oc' — ie. everyone trying to add their favorite change to the list, something very ambitious being started, and never becoming useful. I'd like cork to be useful fast, and that's why my current goal is to make it compile what rock compiles, no more, no less.
- Re generics: we're keeping reified generics, no matter what happens. If you're wondering why, take a look at `format` (or ooc varargs in general), or think about what happens when you compile a library separately and _then_ instanciate its types with type arguments it didn't know about. Or read articles [written by smarter people than me](http://people.cs.uchicago.edu/~jacobm/pubs/templates.html) about the various arguments against templates. I'm well aware of `gc_malloc` / `memcpy`'s problems, but that's how universally-quantified generics should work in ooc still (although gc_malloc could probably be replaced with `alloca` with little problems — my main concern with stack allocation is blowing up the stack). Performance / implementation issues is a separate problem: I'm satisfied with ooc generics are supposed to work, I'm just not satisfied with how rock deals with them internally.

As for arrays, I definitely agree it's one of the things that should be solved in cork - and in fact, rock's test suite contains [a cover template based implementation of Array](https://github.com/fasterthanlime/rock/blob/master/test/compiler/covers/cover-templates.ooc) that could probably be the built-in.

It sounds like I'm asking for feedback then immediately shutting it down, but that's not really what's happening: in the interest of things moving forward (rather than be stuck in "wouldn't it be nice if" land), what I would like to discuss the most is: the various ASTs that should be involved in ooc compilation, what exactly is the responsibility of backends, how many passes we need and what should happen in them, etc.

> Taking tuple for example, unwrapping it to VariableDecl..., BinaryOp..., VariableAccess is clearly a c-based action. If someone is making go or python backend, there's even no need to touch the tuple itself.

Not really, there's plenty of potential target languages that don't have tuples. By unwrapping things like that in the middle-end, we make sure the semantics are those of ooc!

**Comment by alexnask** [comment]
@zhaihj  

I agree templates are really good in certain, if not most, situations, I'm just saying we already do have them, through cover template (I guess we could support class template too).  

Generics are useful because of the runtime type information.  
I think a good example is how VarArgs are implemented, which are "just" alternating a Class and the Object/data itself.  

Of course, VarArgs could be implemented as variadic template arguments instead but that would mean that a new function is created for each different combination of parameters (imagine how many versions of format would be generated :P) or we could "just" use C varargs, although iirc they require the use of C macros that take a C type, which is tricky to implement at the moment to say the least.  

This is just one example, generics are useful in places where you don't really need to do specific operations depending on the type you passed, or you have a short list of types you will match against.  

Of course, I agree the SDK could easily provide a cover template version of an ArrayList but this is something we will have to consider carefully (what should use generics and what should be templated).  

Another idea is specilizing for types used often, although there are some issues with the implementations of containers which would still not make it optimal without some additional features (Amos has a good presentation on that somewhere)

**Comment by fasterthanlime** [comment]
> or we could "just" use C varargs, although iirc they require the use of C macros that take a C type, which is tricky to implement at the moment to say the least.

C varargs carry no type information at all, by the way. That's why printf can occasionally print garbage, if you mess up your format string. But our format validates its arguments:

<img width="939" alt="screen shot 2015-07-30 at 11 44 17" src="https://cloud.githubusercontent.com/assets/7998310/8980372/5ede30ce-36b0-11e5-8f92-5447fc1fa9ec.png">

**Comment by alexnask** [comment]
Yes, I am aware, C varargs are basically a bunch of compiler-specific macros that tell the compiler to get values of some type we expect.  

That is why the format specifiers exist in the first place, we could easily get rid of them if we wanted :)

**Comment by fasterthanlime** [comment]
> Another idea is specilizing for types used often, although there are some issues with the implementations of containers which would still not make it optimal without some additional features (Amos has a good presentation on that somewhere)

Specialization is a perfectly good solution for the 'performance problem of ooc generics' — falling back to simple templates is handicapping the language and it's a lazy solution. (Yes, generics are hard to deal with in compilers, especially when you have no idea how to implement them — in j/ooc it was a nightmare, in rock it's an ordeal, maybe in cork it'll be a pleasure?).

The main thing I mentioned in [my presentation about specialization](https://speakerdeck.com/nddrylliog/generic-specialization-in-ooc) was that we needed some sort of slice assignment primitive to implement things like ArrayList removeAt() properly (but in the meantime, falling back to moving each element individually would work).

It's not easy to implement a new language, and I understand that it's frustrating for its users, but "why not do manual memory management just like C" or "why not do templates just like C++/D" is never an answer I'll accept :) The main problems in ooc are implementation problems, not design problems. The design is different, but not bad in the absolute or requiring the implementation to be slow in any way.

**Comment by horasal** [comment]
Maybe my words are misleading but.. I didn't talk about doing manual memory management.
I meant that since OOC allows `gc=off`, gc-free should be considered during implementing the compiler.

I should also clearify that if generic can be efficient enough, I will be happy to use it. (In fact, I don't really care about the "design" but only the "efficiency". So my stupid question was not criticism on the design of ooc. Template was meantioned just because it's the only solution I can figure out.)

Before getting off topic too far away, I think I should forget these.

In fact I also tried to implement [ooc-array](https://gist.github.com/zhaihj/8575451529f5d8fabbf0), but it ended up with nothing because there're too much array-related code and I didn't know how to start.

**Comment by davidhesselbom** [comment]
> In fact, I don't really care about the "design" but only the "efficiency"

Me too, mostly. If something can be done to make code that doesn't use the GC less cluttered, that would make me happy, but all of what I said about stack allocation was just meant as suggestions. You know much better than I do what needs to be done to get where you want to go, so I'll take my leave of this discussion now :)

> I'd like cork to be useful fast

Right. Best of luck!

**Comment by fasterthanlime** [comment]
> Generics efficiency

Here's an example of what I'd like to see in cork: there's basically 3 ways we can handle manipulating generic values in ooc.

Let's say we have something like:

``` ooc
identity: func <T> (x: T) -> T {
  x
}

nothing: func <T> (x: T) {
 // told you.
}

f: func <T> (x: T) -> T {
  y := identity(x)
  nothing(y)
  y
}
```

Then the `y` variable is a generic value, so, by ooc standard, it should:
- Have by-value semantics
- Be able to hold a value of any size (size of a pointer of classes, underlying type size for covers)

Now, rock will probably generate code like this:

``` c
void f(lang_Class T, uint8_t* x, uint8_t* ret) {
  uint8_t* y = gc_malloc(T->size);
  identity(T, x, y);
  nothing(T, y);
  // ret can be null if the return value is ignored
  if (ret) {
    memcpy(ret, y, T->size);
  }
  return;
}
```

That's way number one - using `gc_malloc`, so that we don't have to worry when it's freed.

For C code readability, let's assume cork is smarter and uses a typedef like this:

``` c
typedef uint8_t* Any;

void f(lang_Class T, Any x, Any ret) {
  Any y = gc_malloc(T->size);
  identity(T, x, y);
  nothing(T, y);
  if (ret) {
    memcpy(ret, y, T->size);
  }
  return;
}
```

That's better. Now, the thing with generic values - in the generated C code - is that they're actually references to something else, right? When you assign two generic values, the lhs is set to point to what the rhs points, but then the rhs can be discarded. So, we never need to allocate on the heap, because the lifetime of what the generic value _refers to_ is separate.

Let's take a real value in C for example:

``` c
void f() {
  // a is on the stack (well, in a register.)
  int a = 42;

  // a has by-value semantics, so its value is *copied* when passed as
  // an argument
  nothing(a);

  // a is freed when returning from f, but it's okay - we haven't
  // passed a dangling pointer to `nothing`, even if it retained
  // the value of a, it got a copy of it.
  return;
}
```

And generics do work that way, so there is only one reason, ever, to want to allocate generic values (we could call them 'boxes', I feel like that's a better term) on the heap, and that's if you have a lot of recursion (call stack grows a lot) and your generic objects are huge structs.

So, cork could generate something like this instead:

``` c
typedef uint8_t* Any;

void f(lang_Class T, Any x, Any ret) {
  Any y = alloca(T->size);
  identity(T, x, y);
  nothing(T, y);
  if (ret) {
    memcpy(ret, y, T->size);
  }
  // after this point, `y` is freed - look, ma, no GC!
  return;
}
```

Or, if your C compiler supports [VLAs](https://en.wikipedia.org/wiki/Variable-length_array), it could generate this:

``` c
typedef uint8_t* Any;

void f(lang_Class T, Any x, Any ret) {
  uint8_t[T->size] tmp;
  Any y = (Any) tmp;
  identity(T, x, y);
  nothing(T, y);
  if (ret) {
    memcpy(ret, y, T->size);
  }
  // same shenanigans
  return;
}
```

Of course, for that to happen, we can't have the middle-end just go away and add some `gc_malloc` calls to variable declarations because they're of a generic type. Instead, we could use a few new AST nodes:

``` c
f: func <T> (x: T) {
  y: T // VariableDecl(name = "y", expr = BoxCreation(T))
  y = x // BoxCopy(Access("y"), Access("x"))
  x as Int  // Unboxing(Access("x"), Int)
}
```

etc. And then, from there, `BoxCreation` could be translated either to `gc_malloc`, to `alloca`, or VLAs.

Note that `BoxCreation` wouldn't necessarily be done in the middle-end, but perhaps in some transitional AST in-between the ooc AST and the C ast. Definitely not in the resolve phase, at any rate.

That's the kind of idea I want to discuss in this thread :)
